### PR TITLE
State pattern is maintainable

### PR DESCRIPTION
**Pull request by ko22009**
I will refactoring this pattern.

### CODE REVIEW

File: Behavioral/State/ContextOrder.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace DesignPatterns\Behavioral\State;
+
+class ContextOrder extends StateOrder
+{
+    public function getState():StateOrder
```
Comment (gmsantos) [inline Behavioral/State/ContextOrder.php]: PHP 7 sintax should go to `php7` branch (https://github.com/domnikl/DesignPatternsPHP/pull/169#issuecomment-248949061). Master branch still PHP 5.6 compatible

File: Behavioral/State/CreateOrder.php [hunk 2]
```diff
@@ -2,34 +2,17 @@
...
+        $this->setStatus('created');
     }
 
-    public function getStatus(): string
+    protected function done()
     {
-        return $this->details['status'];
+        $order = new ShippingOrder();
+        static::$state = $order;
+        //$this->setStatus('shipping');
```
Comment (gmsantos) [inline Behavioral/State/CreateOrder.php]: comment

File: Behavioral/State/ContextOrder.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        static::$state = $state;
+    }
+
+    public function done()
+    {
+        static::$state->done();
+    }
+
+    public function getStatus(): string
```
Comment (gmsantos) [inline Behavioral/State/ContextOrder.php]: php 7 only

File: Behavioral/State/StateOrder.php [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+     * @var StateOrder $state
+     */
+    protected static $state;
+
+    /**
+     * @return mixed
+     */
+    abstract protected function done();
+
+    protected function setStatus(string $status)
```
Comment (gmsantos) [inline Behavioral/State/StateOrder.php]: php 7 only

File: Behavioral/State/Tests/StateTest.php [hunk 5]
```diff
@@ -2,33 +2,30 @@
...
 
     public function testCanCompleteShippedOrder()
     {
-        $order = (new OrderRepository())->findById(2);
-        $order->completeOrder();
+        $order = new CreateOrder();
+        $contextOrder = new ContextOrder();
+        $contextOrder->setState($order);
+        $contextOrder->done();
+        $contextOrder->done();
```
Comment (gmsantos) [inline Behavioral/State/Tests/StateTest.php]: why `done()` is fired twice?

Comment (ko22009) [inline Behavioral/State/Tests/StateTest.php]: It's show, that's order from create go to shipping and after to completed. Our states is connected. It's normal behaviour, one order transform from one state to another.

File: Behavioral/State/ContextOrder.php (hunk 1 above)
Comment (ko22009) [inline Behavioral/State/ContextOrder.php]: What you say about https://github.com/domnikl/DesignPatternsPHP/blob/master/Structural/Adapter/Book.php#L22?

Comment (gmsantos) [inline Behavioral/State/ContextOrder.php]: My bad, didn't know that php7 branch was merged on master

File: Behavioral/State/CreateOrder.php (hunk 2 above)
Comment (domnikl) [inline Behavioral/State/CreateOrder.php]: also `$order` is superfluous here: `static::$state = new ShippingOrder()`

**Comment by domnikl** [comment]
I did, thanks!

**Comment by ko22009** [comment]
Clear.

**Pull request merged**